Add new `envelope` module along with an `envelope::Detector` type

mitchmindtree opened the pull request:
Envelope detection (also known as envelope following) is commonly used
in audio for monitoring, side-chaining, etc.

`Detector` is generic over its `Detect`ion type and also allows for
specifying `attack` and `release`. Implementations of `Detect` are
provided for both RMS and peak detection.

An accompanying `Signal::detect_envelope` method has been added that
produces an adaptor yielding detected envelope frames.

andrewcsmith commented:
Is `Detector` an onset detector? If that's the case, I would prefer a more primitive struct similar to what SuperCollider and other DSP languages have (maybe called `Onset` instead of `Detector`, to make it more specific). SuperCollider's [Onsets](http://doc.sccode.org/Classes/Onsets.html) class just outputs 1 when an onset is detected in the given frame and 0 when no onset is detected. I think that would be a more general application than a signal-modulator like what you've got here...the envelope-follower or whatever can then be another separate `struct` that ties the two together.

mitchmindtree commented:
OK I just rebased onto master and fixed up the commits so it should be easier to read the actual changes introduced by this PR now!

> Is Detector an onset detector?

`envelope::Detector` is an [envelope detector](https://en.wikipedia.org/wiki/Envelope_detector), "used to detect the amplitude variations of an incoming signal to produce a control signal that resembles those variations" - I think [this picture](https://upload.wikimedia.org/wikipedia/commons/4/4d/C_Envelope_follower.png) from the wiki describes the behaviour nicely.

> the envelope-follower or whatever can then be another separate struct that ties the two together

From my (very limited) understanding onset detection can be quite a bit more complicated than plain envelope detection as it attempts to take a more psycho-acoustic approach to determine the perceivable onset of a sound. I'd imagine that it would be more common to use envelope detection along with some other FFT stuff and other heuristics to create an onset detector, rather than the other way around. I could be totally wrong though!

andrewcsmith commented:
Ah, gotcha. Sounds good then, and makes sense to me!
